source/server/connection_handler_impl.cc

```diff
@@ -168,7 +168,17 @@ void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
 
 void ConnectionHandlerImpl::ActiveListener::onAccept(
     Network::ConnectionSocketPtr&& socket, bool hand_off_restored_destination_connections) {
-  Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+  // In case the connection was redirected transparently, set the local
+  // address as restored.
+  if (config_.transparent()) {
+    const Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+    const Network::Address::InstanceConstSharedPtr bind_address = config_.socket().localAddress();
+    if (*local_address != *bind_address) {
+      socket->setLocalAddress(bind_address, false);
+      socket->setLocalAddress(local_address, true);
```

Comment:
Why do we need both? Isn't second line enough?

Reply:
No, because of the assert in `setLocalAddress`: https://github.com/envoyproxy/envoy/blob/master/source/common/network/listen_socket_impl.h#L74
```
ASSERT(!restored || *local_address != *local_address_);
```
The only way to work around this assert is to first set it to another address, then set the address as restored. I just picked the bind address as the other address.

Reply:
Why is the address marked as restored already? That shouldn't be the case.

Also, this condition is most likely a bit too aggressive, since doesn't allow using PROXY protocol along with `SO_ORIGINAL_DST` and/or `IP_TRANSPARENT`... although I'm not sure if that's valid use case.

Reply:
No, what the assert checks is that if we try to mark an address as restored it must be different from the current address.

I can remove the assert if you feel that’s ok.

Reply:
Ooops, I've misread `restored` as `restored_`... I think the `ASSERT()` might be no longer valid, since it clearly doesn't work for `IP_TRANSPARENT`.

Reply:
I realized that `SO_ORIGINAL_DST` has the same behavior for `REDIRECT`ed and `TPROXY`ed connections. So the `OriginalDst` was almost already working for `TPROXY`ed connections. I just needed to remove the `ASSERT` in `setLocalAddress()`. That makes the code simpler.